Fix buildJdbcUrl to apply connection properties

seoyoniee opened the pull request:
<!-- Thank you for submitting a pull request! Find more information
at https://trino.io/development/process.html,
at https://trinodb.github.io/trino-gateway/development/#contributing
and contact us on #trino-gateway-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description
Fixing buildJdbcUrl to apply connection properties


<!-- Provide details that help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
fixing https://github.com/trinodb/trino-gateway/issues/650


<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
* Fix buildJdbcUrl to apply connection properties (https://github.com/trinodb/trino-gateway/issues/650)
```

seoyoniee commented:
@cla-bot check

mosabua commented:
@cla-bot check

mosabua reviewed (commented):
I think this requires a pretty static format for the JDBC URL. Can you confirm that this works for MySQL, Oracle, and PostgreSQL as we suggest. At a minimum I think we should add some simple test to verify that the property you reported as failing now works in terms of passing it through. Overall I am ok with this approach though. Might be worth looking in the Trino codebase if we have any similar setup and tests for it.

seoyoniee commented:
@mosabua 
As you pointed out, the code I modified previously had a pretty static format and wouldn't work well for various database, especially Oracle. So I made two modifications:
1. re-modified buildJdbcUrl to ensure it works correctly even when query parameters contain slashes 083ee644be7a7c35db220b37df2d09590627f632
2. I also added test code for h2, mysql, Oracle, and PostgreSQL 5c0b9dd2e1cea6663f8b3251cf4b4c4861ee8eb6

xkrogen commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,11 +60,14 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
         String jdbcUrl = configuration.getJdbcUrl();
         if (routingGroupDatabase != null) {
-            jdbcUrl = jdbcUrl.substring(0, jdbcUrl.lastIndexOf('/') + 1) + routingGroupDatabase;
+            int queryParamsIndex = jdbcUrl.indexOf("?");
+            String baseUrl = (queryParamsIndex != -1) ? jdbcUrl.substring(0, queryParamsIndex) : jdbcUrl;
+            String queryParams = (queryParamsIndex != -1) ? jdbcUrl.substring(queryParamsIndex) : "";
+            jdbcUrl = baseUrl.substring(0, baseUrl.lastIndexOf('/') + 1) + routingGroupDatabase + queryParams;
```
Instead of manually implementing parsing here, why not convert it into a `java.net.URL` to get the parsing for free?

seoyoniee replied:
@xkrogen 
as you know, the standard URI syntax is in the format of `scheme:[//[userinfo@]host[:port]][path][?query][#fragment]`. Unlike MySQL and PostgreSQL, Oracle's jdbc urls, as show below, differ from this stardard format, making parsing difficult.
```
jdbc:oracle:thin:@<HOST>:<PORT>:<SID>
jdbc:oracle:thin:@//<HOST>:<PORT>/<SERVICE>
```

To use the `java.net.URI` object, it seems necessary to handle Oracle differently from other databases. Instead of manual processing, does this approach seem better?

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,56 @@
+            Arguments.of("jdbc:oracle:thin:@//localhost:1521/mydb?sessionTimeZone=Asia/Seoul", "newdb", "jdbc:oracle:thin:@//localhost:1521/newdb?sessionTimeZone=Asia/Seoul"));
+    }
+
+    @ParameterizedTest
```
Please follow the test guideline: https://trino.io/docs/current/develop/tests.html#avoid-data-providers-and-parametric-tests

seoyoniee replied:
Following the guidelines, I've replaced parametric tests with explicit test cases. https://github.com/trinodb/trino-gateway/pull/651/commits/6f896c131b7f29a231a7c7e93c4d0dad5c911480

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,56 @@
+
+    @ParameterizedTest
+    @MethodSource("provideJdbcUrlAndDatabase")
+    public void testBuildJdbcUrl(String inputJdbcUrl, String database, String expectedJdbcUrl)
```
```suggestion
    void testBuildJdbcUrl(String inputJdbcUrl, String database, String expectedJdbcUrl)
```

seoyoniee replied:
removed `public` from all test code. https://github.com/trinodb/trino-gateway/pull/651/commits/ed96d7fa1af5fb1d8eb0c6403f98883190e86187

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,56 @@
+
+final class TestJdbcConnectionManager
+{
+    JdbcConnectionManager connectionManager;
+    DataStoreConfiguration db = new DataStoreConfiguration("", "sa", "sa", "", 4, true);
```
Move to the test method.

seoyoniee replied:
I moved it to the test method https://github.com/trinodb/trino-gateway/pull/651/commits/4cf534042f8ad00414c674b6ad99df74a0a37fec

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,162 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+final class TestJdbcConnectionManager
```
Please add a helper method and simplify the redundant code.

seoyoniee replied:
added helper method and fixed to use it (https://github.com/trinodb/trino-gateway/pull/651/commits/01de0f90f2f565f0e63fd4619cae45e1868de0cb)

mosabua reviewed (approved):
This looks good to me now. Can you squash commits @seoyoniee ?

Anything else to change @ebyhr ?

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,106 @@
+
+final class TestJdbcConnectionManager
+{
+    private JdbcConnectionManager createConnectionManager(String jdbcUrl)
```
Please move this method to the bottom.  We locate helper method after the caller in this project.

seoyoniee replied:
https://github.com/trinodb/trino-gateway/pull/651/commits/77dd5fea6f5b8f690882e715adc04f712bba93e1

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,11 +63,26 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
     }
 
-    private String buildJdbcUrl(@Nullable String routingGroupDatabase)
+    public String buildJdbcUrl(@Nullable String routingGroupDatabase)
```
The package-private is sufficient, right? Please add `@VisibleForTesting` annotation.

seoyoniee replied:
yes right! https://github.com/trinodb/trino-gateway/pull/651/commits/7d2d2ae98018a77327cf14b175262ccf05365f0c

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,11 +63,26 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
+            }
+            catch (URISyntaxException e) {
+                throw new RuntimeException(e);
+            }
```
Please return `jdbcUrl` early when `routingGroupDatabase` is null so we can avoid redundant nesting: 
```java
        if (routingGroupDatabase == null) {
            return jdbcUrl;
        }

        try {
...
```

seoyoniee replied:
https://github.com/trinodb/trino-gateway/pull/651/commits/ba8c71e4f0072c24f126f8dc8dd6fc6dc72f0a90

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManager.java:
```diff
@@ -0,0 +1,106 @@
+
+final class TestJdbcConnectionManager
+{
+    private JdbcConnectionManager createConnectionManager(String jdbcUrl)
```
This method can be static.

seoyoniee replied:
https://github.com/trinodb/trino-gateway/pull/651/commits/8d099e733f7cff235f664f55c28d0626add03cec

Chaho12 commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,13 +64,29 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
+                    uri.getFragment());
+            return jdbcUrl.substring(0, index) + newUri;
+        }
+        catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Hey @ebyhr WDYT of catching more broder exception like RuntimeException or Exception
Because NPE could happen if jdbcUrl is none, or IndexOutOfBounds if / is not present, or IllegalArgumentException if `Path.of(uri.getPath())` is malformed path. 

```suggestion 
    catch (Exception e) { // catch all
        Logger.get(getClass()).error(e, "Failed to build JDBC URL for routing group '%s'", routingGroupDatabase);
        throw new RuntimeException(e);
    }
```

seoyoniee replied:
Do you think it's better to do catching more brother exception?
The exceptions you mentioned are treated as if statements, so I think it looks good now

Chaho12 commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,13 +64,29 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
+            return jdbcUrl;
+        }
+        try {
+            int index = jdbcUrl.indexOf("/") + 1;
```
Maybe add check?
```suggestion
            int index = jdbcUrl.indexOf("/") + 1;
            if (index < 0) {
                throw new IllegalArgumentException("Invalid JDBC URL: no '/' found in " + jdbcUrl);
            }
```

seoyoniee replied:
I've also fixed the part where index would be 0 if there was no /, as you mentioned. https://github.com/trinodb/trino-gateway/pull/651/commits/a364ee8d78b7053ec61283d43a7f89fc68a45ec8

Also, since that method was a bit long, I've refactored it by extracting the logic that returns the new URI into a separate method called `getUriWithRoutingGroupDatabase`.  https://github.com/trinodb/trino-gateway/pull/651/commits/c850305afeb6a24bf989488b8c3d386c6d040a07

Chaho12 reviewed (commented):
Can you add some failures in test cases? For incorrect jdbc urls?

seoyoniee commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -60,13 +64,29 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
+    @VisibleForTesting
+    String buildJdbcUrl(@Nullable String routingGroupDatabase)
     {
         String jdbcUrl = configuration.getJdbcUrl();
```
https://github.com/trinodb/trino-gateway/pull/651/commits/7e1b2e87a15ee904acf95c0b5a17850275ea7180

mosabua commented:
Thoughts about merging this @Chaho12 @ebyhr ?

Chaho12 reviewed (approved):
LGTM 👍

Chaho12 commented:
Can you sqaush commits to a single one @seoyoniee ?

seoyoniee commented:
> Can you sqaush commits to a single one @seoyoniee ?

@Chaho12 
Thank you for review :) I squash commits to one
